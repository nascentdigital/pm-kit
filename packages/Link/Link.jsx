import React, { forwardRef } from 'react'
import PropTypes from 'prop-types'
import { css, jsx } from '@emotion/core'
import { size, weight } from '@pm-kit/typography'
import { parkGreen } from '@pm-kit/colours'

const linkStyle = css`
  font-size: ${size.bodySmall};
  color: ${parkGreen};
  &::hover {
    color: ${parkGreen};
    font-size: ${size.bodyMedium};
  }
`

const Link = ({ children, forwardedRef, link, ...rest }) => {
  let variantStyles

  return (
    <>
      <a css={linkStyle} href={link} target="_blank" rel="noopener noreferrer">
        {children}
      </a>
    </>
  )
}

const LinkWithRef = forwardRef((props, ref) => <Link {...props} forwardedRef={ref} />)

LinkWithRef.propTypes = {
  /**
   * The content. Can be text, other components, or HTML elements.
   */
  children: PropTypes.node.isRequired,
  /**
   * The external Link
   */
  link: PropTypes.string,
}

LinkWithRef.defaultProps = {
  children: undefined,
  link: undefined,
}

export default LinkWithRef
