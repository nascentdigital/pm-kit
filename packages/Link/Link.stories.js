import React from 'react'
import { withKnobs, text, select } from '@storybook/addon-knobs'
import Link from './Link.jsx'
import { version } from './package.json'

// export default {
//   title: 'Design System|Link',
//   component: Link,
//   parameters: {
//     componentSubtitle: `version ${version}`,
//   },
//   decorators: [withKnobs],
// }

export const Link1 = () => <Link link="http://www.google.com">Go To Google</Link>
