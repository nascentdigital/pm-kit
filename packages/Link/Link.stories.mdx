import { Meta, Story, Preview, Props, Subtitle } from '@storybook/addon-docs/blocks'
import { withKnobs } from '@storybook/addon-knobs'
import Link from './Link.jsx'
import Paragraph from '@pm-kit/paragraph'
import { size, weight } from '@pm-kit/typography'
import { version } from './package.json'

<Meta title="Design System|Link" componenent={Link} />

# Link
<Subtitle>
version {version}
</Subtitle>

## Usage
* To install `Link` package with npm run terminal:

```jsx
npm i @pm-kit/link
```
* Import `Link` in your componenent to render

```jsx
import Link from '@pm-kit/link'
```

## Usage Guidelines

When you use `target="_blank"` with Links, it is recommended to always set `rel="noopener"` or `rel="noreferrer"` when linking to third party content.

* `rel="noopener"` prevents the new page from being able to access the `window.opener` property and ensures it runs in a separate process. 
Without this, the target page can potentially redirect your page to a malicious URL.

+ `rel="noreferrer"` has the same effect, but also prevents the Referer header from being sent to the new page. 

Removing the referrer header will affect analytics.

## Properties 

<Props of={Link} />

## Stories

#### Default
<Preview>
<Story name="Default">
  <Link href="http://www.google.com">Go To Google</Link>
</Story>
</Preview>

#### Bold
<Preview>
<Story name="Bold">
  <Paragraph weight={weight.bold}>
    <Link href="http://www.google.com" target="_blank">Go To Google</Link>
  </Paragraph>
</Story>
</Preview>