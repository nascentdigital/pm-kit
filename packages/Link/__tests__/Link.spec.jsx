import React from 'react'
import { render } from 'enzyme'

import Link from '../Link'

describe('Link', () => {
  it('renders', () => {
    const link = render(<Link link="http://www.google.com">Google</Link>)

    expect(link).toMatchSnapshot()
  })
})
