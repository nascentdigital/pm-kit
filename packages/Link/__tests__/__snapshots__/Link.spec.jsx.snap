// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link load document in a new window 1`] = `
.emotion-0 {
  color: #034045;
}

.emotion-0:hover {
  color: #034045;
}

<a
  class="emotion-0"
  href="http://www.google.com"
  target="_blank"
>
  Google
</a>
`;

exports[`Link load document to full original window 1`] = `
.emotion-0 {
  color: #034045;
}

.emotion-0:hover {
  color: #034045;
}

<a
  class="emotion-0"
  href="http://www.google.com"
  target="_top"
>
  Google
</a>
`;

exports[`Link load document to immediate frameset parent in new tab 1`] = `
.emotion-0 {
  color: #034045;
}

.emotion-0:hover {
  color: #034045;
}

<a
  class="emotion-0"
  href="http://www.google.com"
  target="_parent"
>
  Google
</a>
`;

exports[`Link renders 1`] = `
.emotion-0 {
  color: #034045;
}

.emotion-0:hover {
  color: #034045;
}

<a
  class="emotion-0"
  href="http://www.google.com"
  target="_self"
>
  Google
</a>
`;
